plugins/out_s3/s3_parquet.c

```diff
@@ -0,0 +1,500 @@
+    if (ret != 0) {
+        flb_plg_error(ctx->ins, "failed to initialize parquet batch mutex");
+        return -1;
+    }
+
+    flb_plg_debug(ctx->ins,
+                 "parquet batch manager initialized: batch_size=%zu, timeout=%ds",
+                 ctx->file_size, (int)ctx->upload_timeout);
+
+    return 0;
+}
+
+/* Flush all pending Parquet batches before shutdown */
+static int parquet_batch_flush_all(struct flb_s3 *ctx)
+{
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct parquet_batch *batch;
+    struct multipart_upload *m_upload;
+    int ret;
+    int flushed_count = 0;
+    int error_count = 0;
+
+    flb_plg_info(ctx->ins, "Flushing all pending parquet batches before shutdown");
+
+    pthread_mutex_lock(&ctx->parquet_batch_lock);
+
+    mk_list_foreach_safe(head, tmp, &ctx->parquet_batches) {
+        batch = mk_list_entry(head, struct parquet_batch, _head);
+
+        /* Skip empty batches */
+        if (!batch->chunk || batch->accumulated_size == 0) {
+            flb_plg_debug(ctx->ins,
+                         "Skipping empty parquet batch for tag '%s'",
+                         batch->tag);
+
+            /* Clean up the batch structure even if skipping */
+            mk_list_del(&batch->_head);
+            if (batch->chunk) {
+                s3_store_file_delete(ctx, batch->chunk);
+                batch->chunk = NULL;
+            }
+            if (batch->tag) {
+                flb_free(batch->tag);
+            }
+            flb_free(batch);
+            continue;
+        }
+
+        flb_plg_info(ctx->ins,
+                    "Flushing parquet batch for tag '%s': %zu bytes, %d appends",
+                    batch->tag, batch->accumulated_size, batch->append_count);
+
+        /* Get or create multipart upload for this tag */
+        m_upload = get_upload(ctx, batch->tag, batch->tag_len);
+        if (!m_upload) {
+            m_upload = create_upload(ctx, batch->tag, batch->tag_len,
+                                    batch->create_time);
+            if (!m_upload) {
+                flb_plg_error(ctx->ins,
+                             "Failed to create upload for parquet batch '%s', "
+                             "data will be discarded",
+                             batch->tag);
+                error_count++;
+                continue;
+            }
+        }
+
+        /* Temporarily unlock to allow conversion and upload */
+        pthread_mutex_unlock(&ctx->parquet_batch_lock);
+
+        /* Convert and upload the batch */
+        ret = parquet_batch_convert_and_upload(ctx, batch, m_upload);
+
+        /* Re-lock for next iteration */
+        pthread_mutex_lock(&ctx->parquet_batch_lock);
+
+        if (ret < 0) {
+            flb_plg_error(ctx->ins,
+                         "Failed to flush parquet batch for tag '%s', "
+                         "data will be discarded",
+                         batch->tag);
+            error_count++;
+        } else {
```

Comment:
We need a break this line like as:

```suggestion
        } 
        else {
```